ykrt/src/deopt.rs

```diff
@@ -268,18 +277,82 @@ unsafe extern "C" fn __ykrt_deopt(
     guardid: usize,
     // Return address of deoptimize call. Used to find the correct stackmap record.
     retaddr: usize,
+    // The parent trace's call stack at the time of the guard failure. Required to assemble
+    // a side trace.
+    jitcallstack: *const c_void,
     // Current stack pointer. Needed to read spilled register values from the stack.
     rsp: *const c_void,
 ) -> NewFramesInfo {
     // Put the CompiledTrace back into an Arc, so it is dropped properly.
     let ctr = Arc::from_raw(ctr);
 
+    // Check if we have a side trace and execute it.
+    if guardid != SIDETRACE_LAST_GUARD_ID {
+        let guard = &ctr.guards[guardid];
+        if let Some(st) = guard.getct() {
+            let registers = Registers::from_ptr(rsp);
+            let live_vars = ctr.smap.get(&retaddr.try_into().unwrap()).unwrap();
+            let mut ykctrlpvars = Vec::new();
+            for (_i, locs) in live_vars.iter().skip(1).enumerate() {
+                assert!(locs.len() == 1);
+                let l = locs.get(0).unwrap();
+                match l {
+                    SMLocation::Register(reg, _size, _off, _extra) => {
+                        let _val = unsafe { registers.get(*reg) };
+                        todo!();
+                    }
+                    SMLocation::Direct(reg, off, _size) => {
+                        // When using `llvm.experimental.deoptimize` then direct locations should always be
+                        // in relation to RBP.
+                        assert_eq!(*reg, 6);
+                        let addr = unsafe { registers.get(*reg) as *mut u8 };
+                        let addr = unsafe { addr.offset(isize::try_from(*off).unwrap()) };
+                        ykctrlpvars.push(addr as u64);
+                    }
+                    SMLocation::Indirect(reg, off, size) => {
+                        let addr = unsafe { registers.get(*reg) as *mut u8 };
+                        let addr = unsafe { addr.offset(isize::try_from(*off).unwrap()) };
+                        let v = match *size {
+                            1 => unsafe { ptr::read::<u8>(addr) as u64 },
+                            2 => unsafe { ptr::read::<u16>(addr as *mut u16) as u64 },
+                            4 => unsafe { ptr::read::<u32>(addr as *mut u32) as u64 },
+                            8 => unsafe { ptr::read::<u64>(addr as *mut u64) },
+                            _ => unreachable!(),
+                        };
+                        ykctrlpvars.push(v);
+                    }
+                    SMLocation::Constant(v) => {
+                        ykctrlpvars.push(*v as u64);
```

Comment:
Can we make this a platform independent size?

Reply:
No, this is always a u64 in LLVM stackmaps.

Reply:
Aha.